sdk/typescript/src/multisig/publickey.ts

```diff
@@ -0,0 +1,289 @@
+import { bcs } from '../types/sui-bcs.js';
+
+type CompressedSignature =
+	| { ED25519: number[] }
+	| { Secp256k1: number[] }
+	| { Secp256r1: number[] };
+
+type PublicKeyEnum = { ED25519: number[] } | { Secp256k1: number[] } | { Secp256r1: number[] };
+
+type PubkeyEnumWeightPair = {
+	pubKey: PublicKeyEnum;
+	weight: number;
+};
+
+type MultiSigPublicKeyStruct = {
+	pk_map: PubkeyEnumWeightPair[];
+	threshold: number;
+};
+
+export type MultiSigStruct = {
+	sigs: CompressedSignature[];
+	bitmap: number;
+	multisig_pk: MultiSigPublicKeyStruct;
+};
+
+type ParsedPartialMultiSigSignature = {
+	signatureScheme: SignatureScheme;
+	signature: Uint8Array;
+	publicKey: PublicKey;
+	weight: number;
+};
+
+export const MAX_SIGNER_IN_MULTISIG = 10;
+
+/**
+ * A MultiSig public key
+ */
+export class MultiSigPublicKey extends PublicKey {
+	private rawBytes: Uint8Array;
+	private multisigPublicKey: MultiSigPublicKeyStruct;
+	private publicKeys: {
+		weight: number;
+		publicKey: PublicKey;
+	}[];
+	/**
+	 * Create a new MultiSigPublicKey object
+	 * @param value MultiSig public key as buffer or base-64 encoded string
+	 */
+	constructor(value: string | Uint8Array | MultiSigPublicKeyStruct) {
+		super();
+
+		if (typeof value === 'string') {
+			this.rawBytes = fromB64(value);
+			this.multisigPublicKey = builder.de('MultiSigPublicKey', this.rawBytes);
+		} else if (value instanceof Uint8Array) {
+			this.rawBytes = value;
+			this.multisigPublicKey = builder.de('MultiSigPublicKey', this.rawBytes);
+		} else {
+			this.multisigPublicKey = value;
+			this.rawBytes = builder.ser('MultiSigPublicKey', value).toBytes();
+		}
+
+		this.publicKeys = this.multisigPublicKey.pk_map.map(({ pubKey, weight }) => {
+			const [scheme, bytes] = Object.entries(pubKey)[0] as [SignatureScheme, number[]];
+			return {
+				publicKey: publicKeyFromBytes(scheme, Uint8Array.from(bytes)),
+				weight,
+			};
+		});
+
+		if (this.publicKeys.length > MAX_SIGNER_IN_MULTISIG) {
+			throw new Error(`Max number of signers in a multisig is ${MAX_SIGNER_IN_MULTISIG}`);
+		}
+	}
+
+	static fromPublicKeys({
+		threshold,
+		publicKeys,
+	}: {
+		threshold: number;
+		publicKeys: { publicKey: PublicKey; weight: number }[];
+	}) {
+		return new MultiSigPublicKey({
+			pk_map: publicKeys.map(({ publicKey, weight }) => {
+				const scheme = SIGNATURE_FLAG_TO_SCHEME[publicKey.flag() as SignatureFlag];
+
+				return {
+					pubKey: { [scheme]: Array.from(publicKey.toRawBytes()) } as PublicKeyEnum,
+					weight,
+				};
+			}),
+			threshold,
+		});
+	}
+
+	/**
+	 * Checks if two MultiSig public keys are equal
+	 */
+	override equals(publicKey: MultiSigPublicKey): boolean {
+		return super.equals(publicKey);
+	}
+
+	/**
+	 * Return the byte array representation of the MultiSig public key
+	 */
+	toRawBytes(): Uint8Array {
+		return this.rawBytes;
+	}
+
+	/**
+	 * Return the Sui address associated with this MultiSig public key
+	 */
+	toSuiAddress(): string {
```

Comment:
Ah this is why this method can't be fully generic. Although you could just override here if you wanted to.